File: DSA/src/Questions/DP/BuySellStocks/KTransactions.java
package Questions.DP.BuySellStocks;

import java.util.Arrays;

public class KTransactions {

    static int getAns(int[] Arr, int n, int ind, int buy, int cap, int[][][] dp) {
        // Base case: If we have processed all stocks or have no capital left, return 0
        // profit
        if (ind == n || cap == 0)
            return 0;

        // If the result for this state is already calculated, return it
        if (dp[ind][buy][cap] != -1)
            return dp[ind][buy][cap];

        int profit = 0;

        if (buy == 0) { // We can buy the stock
            profit = Math.max(0 + getAns(Arr, n, ind + 1, 0, cap, dp),
                    -Arr[ind] + getAns(Arr, n, ind + 1, 1, cap, dp));
        }

        if (buy == 1) { // We can sell the stock
            profit = Math.max(0 + getAns(Arr, n, ind + 1, 1, cap, dp),
                    Arr[ind] + getAns(Arr, n, ind + 1, 0, cap - 1, dp));
        }

        // Store the calculated profit in the dp array and return it
        return dp[ind][buy][cap] = profit;
    }

    static int maxProfitRecursive(int[] prices, int k) {
        int n = prices.length;

        // Creating a 3D dp array of size [n][2][3]

        int[][][] dp = new int[n][2][k + 1];

        // Initialize the dp array with -1
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < 2; j++) {
                Arrays.fill(dp[i][j], -1);
            }
        }

        // Calculate and return the maximum profit
        return getAns(prices, n, 0, 0, k, dp);
    }
    // Time Complexity: O(N*K*3)

    // Reason: There are N*K*3 states therefore at max ‘N*2*3’ new problems will be
    // solved.

    // Space Complexity: O(N*K*3) + O(N)

    // Reason: We are using a recursion stack space(O(N)) and a 3D array (
    // O(N*K*3)).

    static int maxProfitTabulation(int[] prices, int k) {
        int n = prices.length;

        // Creating a 3D dp array of size [n+1][2][3] initialized to 0
        int[][][] dp = new int[n + 1][2][k+1];
        // Each Cell dp[i][j][k] talk about the maximum profit that can be made by
        // performing j transactions till ith day with k limits in hand.

        // Loop through the dp array, starting from the second last stock (ind=n-1)
        for (int ind = n - 1; ind >= 0; ind--) {
            for (int buy = 0; buy <= 1; buy++) {
                for (int cap = 1; cap <= k; cap++) {

                    if (buy == 0) { // We can buy the stock
                        dp[ind][buy][cap] = Math.max(0 + dp[ind + 1][0][cap],
                                -prices[ind] + dp[ind + 1][1][cap]);
                    }

                    if (buy == 1) { // We can sell the stock
                        dp[ind][buy][cap] = Math.max(0 + dp[ind + 1][1][cap],
                                prices[ind] + dp[ind + 1][0][cap - 1]);
                    }
                }
            }
        }

        // The maximum profit with 2 transactions is stored in dp[0][0][2]
        return dp[0][0][k];
    }
    // Time Complexity: O(N*k*3)

    // Reason: There are three nested loops that account for O(N*k*3) complexity.

    // Space Complexity: O(N*k*3)

    // Reason: We are using an external array of size ‘N*k*3’. Stack Space is
    // eliminated.

    static int maxProfitOptimized(int[] prices, int k) {
        int n = prices.length;

        // Create a 2D array 'ahead' and 'cur' to store profit values
        int[][] ahead = new int[3][k + 1];
        int[][] cur = new int[3][k + 1 ];

        // Loop through the prices array, starting from the second last stock (ind=n-1)
        for (int ind = n - 1; ind >= 0; ind--) {
            for (int buy = 0; buy <= 1; buy++) {
                for (int cap = 1; cap <= k; cap++) {

                    if (buy == 0) { // We can buy the stock
                        cur[buy][cap] = Math.max(0 + ahead[0][cap],
                                -prices[ind] + ahead[1][cap]);
                    }

                    if (buy == 1) { // We can sell the stock
                        cur[buy][cap] = Math.max(0 + ahead[1][cap],
                                prices[ind] + ahead[0][cap - 1]);
                    }
                }
            }

            // Copy the values of 'cur' to 'ahead'
            for (int i = 0; i < 3; i++) {
                for (int j = 0; j <= k; j++) {
                    ahead[i][j] = cur[i][j];
                }
            }
        }

        // The maximum profit with 2 transactions is stored in ahead[0][2]
        return ahead[0][k];
    }
    // Time Complexity: O(N*k*3)

    // Reason: There are three nested loops that account for O(N*k*3) complexity

    // Space Complexity: O(1)

    // Reason: We are using two external arrays of size ‘k*3’.

    public static void main(String[] args) {
        int[] prices = {3, 3, 5, 0, 0, 3, 1, 4};
        int k = 5;
        System.out.println(maxProfitRecursive(prices, k));
        System.out.println(maxProfitTabulation(prices, k));
        System.out.println(maxProfitOptimized(prices, k));
    }
}

File: DSA/Python/Questions/DP/BuySellStocks/k_transactions.py
"""
# Question: Best Time to Buy and Sell Stock IV (k transactions)
# Link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/

# Problem Statement:
# Given an array prices[] representing stock prices and an integer k, find the maximum 
# profit that can be earned by doing at most k transactions. One transaction consists
# of buying a stock and then selling it.

# Example:
# Input: prices = [3,2,6,5,0,3], k = 2
# Output: 7
# Explanation: Buy on day 2 (price = 2) and sell on day 3 (price = 6), profit = 6-2 = 4.
# Then buy on day 5 (price = 0) and sell on day 6 (prcice = 3), profit = 3-0 = 3.
"""

from typing import List

# 1. Recursive Solution with Memoization
"""
Algorithm:
1. Use 3D DP array to store states: [index][buy/sell][transactions]
2. For each state:
   - If can buy: choose between buying current stock or skipping
   - If can sell: choose between selling current stock or skipping
3. Track remaining transactions and return maximum profit

Time Complexity: O(n*k*2) - three changing parameters in states
Space Complexity: O(n*k*2) + O(n) - DP array + recursion stack
"""


def max_profit_recursive(self, prices: List[int], k: int) -> int:
    n = len(prices)
    dp = [[[-1] * (k + 1) for _ in range(2)] for _ in range(n)]

    def solve(ind: int, can_buy: int, cap: int) -> int:
        if ind == n or cap == 0:
            return 0

        if dp[ind][can_buy][cap] != -1:
            return dp[ind][can_buy][cap]

        profit = 0
        if can_buy:
            profit = max(-prices[ind] + solve(ind + 1, 0, cap), solve(ind + 1, 1, cap))
        else:
            profit = max(
                prices[ind] + solve(ind + 1, 1, cap - 1), solve(ind + 1, 0, cap)
            )

        dp[ind][can_buy][cap] = profit
        return profit

    return solve(0, 1, k)


# 2. Tabulation Solution
"""
Algorithm:
1. Create 3D DP table: [n+1][2][k+1]
2. Fill table bottom-up starting from last day
3. For each state, calculate maximum profit using:
   - Buy state: max(-price[i] + next_day[sell], skip)
   - Sell state: max(price[i] + next_day[buy], skip)

Time Complexity: O(n*k*2) - three nested loops
Space Complexity: O(n*k*2) - 3D DP array
"""


def max_profit_tabulation(self, prices: List[int], k: int) -> int:
    n = len(prices)
    dp = [[[0] * (k + 1) for _ in range(2)] for _ in range(n + 1)]

    for ind in range(n - 1, -1, -1):
        for can_buy in range(2):
            for cap in range(1, k + 1):
                if can_buy:
                    dp[ind][can_buy][cap] = max(
                        -prices[ind] + dp[ind + 1][0][cap], dp[ind + 1][1][cap]
                    )
                else:
                    dp[ind][can_buy][cap] = max(
                        prices[ind] + dp[ind + 1][1][cap - 1], dp[ind + 1][0][cap]
                    )

    return dp[0][1][k]


# 3. Space Optimized Solution
"""
Algorithm:
1. Use two 2D arrays (ahead and curr) instead of 3D array
2. Process each day and update arrays:
   - Calculate current day profits using ahead array
   - Copy curr to ahead for next iteration
3. Space optimization by using only previous day's values

Time Complexity: O(n*k*2) - three nested loops
Space Complexity: O(k*2) - two 2D arrays
"""


def max_profit_optimized(self, prices: List[int], k: int) -> int:
    n = len(prices)
    ahead = [[0] * (k + 1) for _ in range(2)]
    curr = [[0] * (k + 1) for _ in range(2)]

    for ind in range(n - 1, -1, -1):
        for can_buy in range(2):
            for cap in range(1, k + 1):
                if can_buy:
                    curr[can_buy][cap] = max(
                        -prices[ind] + ahead[0][cap], ahead[1][cap]
                    )
                else:
                    curr[can_buy][cap] = max(
                        prices[ind] + ahead[1][cap - 1], ahead[0][cap]
                    )

        ahead = [row[:] for row in curr]

    return ahead[1][k]


def main():
    solver = KTransactions()
    test_cases = [
        ([3, 3, 5, 0, 0, 3, 1, 4], 2),  # Regular case
        ([1, 2, 3, 4, 5], 2),  # Strictly increasing
        ([7, 6, 4, 3, 1], 2),  # Strictly decreasing
        ([3, 2, 6, 5, 0, 3], 2),  # Multiple peaks
        ([1], 1),  # Single element
        ([1, 2], 1),  # Two elements
    ]

    for i, (prices, k) in enumerate(test_cases):
        print(f"\nTest Case {i + 1}: prices = {prices}, k = {k}")
        print(f"Recursive Solution: {solver.max_profit_recursive(prices, k)}")
        print(f"Tabulation Solution: {solver.max_profit_tabulation(prices, k)}")
        print(f"Optimized Solution: {solver.max_profit_optimized(prices, k)}")


if __name__ == "__main__":
    main()
